### PR TITLE
Push containers to GitHub container registry

### LINKED ISSUE #79323: [build] push containers used for PyTorch build in CI to GitHub container registry
### 🚀 The feature, motivation and pitch

When trying to reproduce build / test problems from CI locally, one of the challenges it to get the same image that is used in CI worker. At the moment the images are stored in private registry. Although one could re-build them locally, this is both time-consuming and potentially error prone. It would be much easier to grab container with the same sha256 as the CI one and run the code inside of it.

cc @malfet 

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by DanilBaibak**
Fixes #79323

### CODE REVIEW

File: .circleci/docker/build_docker.sh [hunk 1]
```diff
@@ -49,6 +50,13 @@ fi
 # Only push if `DOCKER_SKIP_PUSH` = false
 if [ "${DOCKER_SKIP_PUSH:-true}" = "false" ]; then
   docker push "${image}:${tag}"
+
+  if [ "${BUILD_TYPE}" = "release" ]; then
```
Comment (seemethere) [inline .circleci/docker/build_docker.sh:54]: Where do we expect this to get populated?

As well we should probably add a default for this like so:

```suggestion
  if [ "${BUILD_TYPE:-}" = "release" ]; then
```

Comment (seemethere) [reply .circleci/docker/build_docker.sh:54]: I think we can probably probably do something similar to what we do with our `pytorch/builder` workflows where we set an env variable like so (probably name it something like `PUSH_GHCR_IMAGE`):

https://github.com/pytorch/builder/blob/4c1b45bbc550e4b6d45853627fe53cf64b1de71b/.github/workflows/build-conda-images.yml#L21

Comment (DanilBaibak) [reply .circleci/docker/build_docker.sh]: I added the env variable that can be set as the action's input.

File: .circleci/docker/build_docker.sh [hunk 2]
```diff
@@ -49,6 +50,13 @@ fi
 # Only push if `DOCKER_SKIP_PUSH` = false
 if [ "${DOCKER_SKIP_PUSH:-true}" = "false" ]; then
   docker push "${image}:${tag}"
+
+  if [ "${PUSH_GHCR_IMAGE:-}" ]; then
+    # Push docker image to the ghcr.io
+    echo $GHCR_PAT | docker login ghcr.io -u pytorch --password-stdin
```
Comment (seemethere) [inline .circleci/docker/build_docker.sh:56]: GHCR_PAT would be blank on forked pull requests, we might need to pass it from the top level workflow

File: .github/workflows/docker-builds.yml [hunk 3]
```diff
@@ -62,11 +62,14 @@ jobs:
...
         id: build-docker-image
         uses: ./.github/actions/calculate-docker-image
+        env:
+          GHCR_PAT: ${{ secrets.GHCR_PAT }}
         with:
           docker-image-name: ${{ matrix.docker-image-name }}
           always-rebuild: true
           skip_push: false
           force_push: true
+          push-ghcr-image: true
```
Comment (seemethere) [inline .github/workflows/docker-builds.yml:72]: This should only be set true if the event that triggered this is a `push` event

File: .circleci/docker/build_docker.sh [hunk 4]
```diff
@@ -54,6 +55,14 @@ if [ "${DOCKER_SKIP_PUSH:-true}" = "false" ]; then
   if ! docker manifest inspect "${image}:${tag}" >/dev/null 2>/dev/null; then
     docker push "${image}:${tag}"
   fi
+
+  if [ "${PUSH_GHCR_IMAGE:-}" ]; then
```
Comment (seemethere) [inline .circleci/docker/build_docker.sh:59]:
```suggestion
  if [ "${PUSH_GHCR_IMAGE:-}" = "true" ]; then
```

File: .github/workflows/docker-builds.yml (hunk 3 above)
Comment (seemethere) [reply .github/workflows/docker-builds.yml:72]:
```suggestion
          push-ghcr-image: ${{ github.event_name == 'push' }}
```

**Pull request closed without merging**